### ISSUE TITLE
Attribute and analyzer for "soft abstract"

### ISSUE THREAD

**Issue by terrajobst** [dotnet/runtime#38238]
For abstract base types, we sometimes find ourselves in a situation where we'd like to add a new core concept to a type. In order to stay ABI-compatible, we can't add new `abstract` members as this means derived types that were compiled against the earlier version are missing the override, thus causing a `TypeLoadException`.

The general pattern for this that we make the new members `virtual`, ideally with a workable implementation. This doesn't always work, so sometimes the only option is adding an implementation that throws `NotImplementedException`. A good example of this problem is in https://github.com/dotnet/runtime/issues/31362#issuecomment-643438552.

A major downside of this approach is that developers deriving from the base type don't necessarily know that they are supposed to override these new virtual methods to get the best behavior.

One option to solve this problem is by adding a new attribute and analyzer that warns when deriving from it without overriding the marked members.

### API Proposal

```C#
namespace System.Diagnostics.CodeAnalysis
{
    [AttributeUsage(AttributeTargets.Event |
                    AttributeTargets.Method |
                    AttributeTargets.Property)]
    public sealed class ShouldBeOverridenAttribute : Attribute
    {
        public ShouldBeOverridenAttribute();
    }
}
```

### Analyzer Proposal

Let's take [this example from the BCL](https://github.com/dotnet/runtime/issues/31362#issuecomment-643438552) and assume we marked the new virtual members as `[ShouldBeOverriden]` like this:

```C#
namespace System.Threading.Channels
{
    public abstract partial class ChannelReader<T>
    {
        [ShouldBeOverriden]
        public virtual bool CanPeek => false;

        [ShouldBeOverriden]
        public virtual bool TryPeek([MaybeNullWhen(false)] out T item)
        {
            item = default!;
            return false;
        }
    }
}
```

A customer deriving from `ChannelReader<T>` would only override the `abstract` members by default, resulting in this:

```C#
public class MyChannelReader<T> : ChannelReader<T>
             ~~~~~~~~~~~~~~~
{
    public override bool TryRead([MaybeNullWhen(false)] out T item)
    {
        throw new NotImplementedException();
    }

    public override ValueTask<bool> WaitToReadAsync(CancellationToken cancellationToken = default)
    {
        throw new NotImplementedException();
    }
}
```

This would product warnings on the type delaration:

```text
WARNING: ChannelReader<T>.CanPeek should be overriden
WARNING: ChannelReader<T>.TryPeek(out T item) should be overriden
```

### IDE

We may want to change the behavior of the IDE so that the code fixer that generates overrides for abstract members automatically also generates overrides for any members marked as `[ShouldBeOverriden]`.

/cc @dotnet/fxdc @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
I agree this should be an analyzer.  I agree that teh IDE should respect it in features like "implement abstract class".  We would also likely want to hook that up automatically for that new diagnostic id.

Proposal as it stands sounds very reasonable to me.  I would certainly have liked it many times in roslyn itself :)

**Comment by marek-safar** [comment]
I think this should better be fixed at the language level

**Comment by tannergooding** [comment]
That would also require it be solved for F#, VB, etc.

**Comment by jaredpar** [comment]
> That would also require it be solved for F#, VB, etc.

Doing the support in an analyzer or the compiler requires it to be solved in all three languages separately.

> I think this should better be fixed at the C# language level

This is borderline for me. On first reading this feels a bit domain specific in which case it would be better served as an analyzer. 

At the same time though the libraries team has to make a decision: what level of enforcement do they want for this attribute? If they want this actually enforced through the ecosystem then they need to consider the impact on the VB and F# teams as well. That is generally the standard to which APIs in the runtime are held too. If they do want it at that level then we probably end up with a compiler feature because that is the only real avenue available to the F# team.

**Comment by terrajobst** [comment]
* We like the analzyer and we're OK with that not being handled by all languages
    - @terrajobst to fill in more details after the meeting
* The analyzer should only consider missing overrides in the direct base type. Warning on more derived base types isn't useful because the base type is most likely the one that would need to provide behavior.
* We should do a scan for all types that have virtuals that throw
* We should think about how and if we differentiate "this always throw", "this is an optional feature", "this a less performant implementation, please provide a better one" so that we can differentiate the diagnostic ID/severity accordingly.

**Comment by tommcdon** [comment]
@terrajobst is this on track for 5.0, or should we move it to .NET 6?
